adr/ADR-43.md

```diff
@@ -0,0 +1,85 @@
+ * Server-applied limits can result in tombstones with a short per message TTL so that consumers can be notified of limits being processed. Useful in KV watch scenarios being notified about TTL removals
+ * A stream may have a general MaxAge but some messages may have infinite retention, think a schema or type hints in a KV bucket that is forever while general keys have TTLs
+
+Related issues [#3268](https://github.com/nats-io/nats-server/issues/3268)
+
+## Per-Message TTL
+
+We will allow a message to supply a TTL using a header called `Nats-TTL` followed by the duration as seconds.
+
+The duration will be used by the server to calculate the deadline for removing the message based on its Stream 
+timestamp and the stated duration.
+
+The TTL may not exceed the Stream MaxAge to give administrators a hard cap for the purposes of managing resources. 
+The shortest allowed TTL would be 1 second.
+
+A TTL of zero will be ignored, any other unparsable value will result in a error reported in the Pub Ack and the message
+being discarded.
+
+Setting the header `Nats-No-Expire` to `1` will result in a message that will never be expired.
+
+## Limit Tombstones
+
+Several scenarios for server-created tombstones can be imagined, the most often requested one though is when MaxAge
+removes last value (ie. the current value) for a Key.
+
+In this case when the server removes a message and the message is the last in the subject it would place a message 
+with a TTL matching the Stream configuration value.  The following headers would be placed:
+
+```
+Nats-Applied-Limit: MaxAge
+Nats-TTL: 1
+```
+
+The `Nats-Limit-Applied` field is there to support future expansion of this feature.
+
+This behaviour is off by default unless opted in on the Stream Configuration.
+
+## Publish Acknowledgements
+
+We could optionally extend the `PubAck` as follows:
+
+```golang
+type PubAck struct {
+	MsgTTL    uint64 `json:"msg_ttl,omitempty"`
+}
+```
+
+This gives clients a chance to confirm, without Stream Info or should the Stream be edited after Info, if the TTL 
+got applied.
+
+## Stream Configuration
+
+Weather or not a stream support this behavior should be a configuration opt-in. We want clients to definitely know 
+when this is supported which the opt-in approach with a boolean on the configuration would make clear.
+
+We have to assume someone will want to create a replication topology where at some point in the topology these tombstone
+type messages are retained for an audit trail. So a Stream with this feature enabled can replicate to one with it 
+disabled and all the messages that would have been TTLed will be retained.
+
+```golang
+type StreamConfig struct {
+	// AllowMsgTTL allows header initiated per-message TTLs
+	AllowMsgTTL bool          `json:"allow_msg_ttl"`
+
+	// LimitsTTL activates writing of messages when limits are applied with a specific TTL
```

Comment:
So this is another configurable threshold below MaxAge?

Also - consider having setting for disabling `Nats-No-Expire`.

Reply:
When the server places a tombstone after removing the last message for a subject you would need a TTL ok that tombstone but what is the right setting? I think MaxAge but had some disagreement around that so this setting controls the TTL of MaxAge tombstones.

Reply:
Disabling is a good idea but let’s leave it for later after feedback to get the core behaviour out.

Reply:
@ripienaar Yes, it makes sense for tombstone to not live as long as MaxAge. I understant you name it in a way generic enough that we can reuse it for other purposes, however in this case, I wonder if its not better to just call it `TombstoneTTL`. Though I'm not convinced myself.

Reply:
We have client added tombstones and server ones and maybe eventually server ones for multiple reasons - so these are the ones added by the stream limits.